src/tensors/cpu/tensor_operators.cpp

```diff
@@ -654,40 +654,63 @@ void PasteCols(Tensor out_,
-             const Tensor in,
-             const Tensor indices) {
-
-  std::cerr << indices->debug() << std::endl;
-
-  matchOrAbort<IndexType>(indices->type());
-
-  functional::Shape outShape = out->shape();
-  functional::Shape inShape = in->shape();
-
-  auto idxData = indices->data<IndexType>();
-  auto odata = out->data();
-  const auto idata = in->data();
-
-  int size = outShape[3];
-
-  for(int k = 0; k < outShape[0]; ++k) {
-    for(int j = 0; j < outShape[1]; ++j) {
-      int outOffset = k * j * outShape[2] * size + j * outShape[2] * size;
-      int inOffset = k * j * inShape[2] * size + j * inShape[2] * size;
-      for(int i = 0; i < outShape[2]; ++i) {
-        auto idx = idxData[i];
-        int outIndex = outOffset +   i * size;
-        int inIndex  = inOffset  + idx * size;
-        std::copy(idata + inIndex, idata + inIndex + size, odata + outIndex);
+/* Recursive template to implement LoopBeforeAxis. */
+template <class Backend, int Before> struct LoopBeforeAxisImpl {
+  static inline void Loop(
+      const functional::Shape &outShape, int outBase,
+      const functional::Shape &inShape, int inBase,
+      const functional::Shape &idxShape, int idxBase,
+      int axisCPU,
+      Backend backend) {
+    // Loop over this dimension.
+    const int dim = axisCPU - Before;
+    if (dim < 0) {
+      // This template is instantiated for every possible dimension, typically
+      // more than before the axis.
+      LoopBeforeAxisImpl<Backend, Before - 1>::Loop(outShape, outBase, inShape, inBase, idxShape, idxBase, axisCPU, backend);
+    } else {
+      const int outStride = outShape.stride(dim);
+      const int end = outShape.dim(dim);
+      const int inStride = inShape.stride(dim);
+      const int idxStride = idxShape.bstride(dim);
+      for (int i = 0; i < end; ++i) {
+        LoopBeforeAxisImpl<Backend, Before - 1>::Loop(outShape, outBase, inShape, inBase, idxShape, idxBase, axisCPU, backend);
+        outBase += outStride;
+        inBase += inStride;
+        idxBase += idxStride;
       }
     }
   }
+};
+
+/* We're at the axis, call the functor. */
+template <class Backend> struct LoopBeforeAxisImpl<Backend, 0> {
+  static inline void Loop(
+      const functional::Shape &, int outBase,
+      const functional::Shape &, int inBase,
+      const functional::Shape &, int idxBase,
+      int /*axisCPU*/,
+      Backend backend) {
+    backend(outBase, inBase, idxBase);
+  }
+};
+
+/* Jointly loop over dimensions [0, axisCPU) of three tensors out, in, and
+ * indices.  Call the Backend functor for each iteration of the loop.
+ * Backend is a functor taking the tensors and base indices into them:
+ * Backend::operator()(
+ *   int out_base,
+ *   int in_base,
+ *   int indices_base);
+ */
+template <class Backend> inline void LoopBeforeAxis(
+    const functional::Shape &outShape,
+    const functional::Shape &inShape,
+    const functional::Shape &idxShape,
+    int axisCPU,
+    Backend backend) {
```

Comment:
Let's use a different name than `Backend`. We use backend for CPU/GPU device things, had me confused for a sec.